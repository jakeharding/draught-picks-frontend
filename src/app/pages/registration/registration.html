<!--
  Generated template for the RegistrationPage page.

  See http://ionicframework.com/docs/components/#navigation for more info on
  Ionic pages and navigation.
-->
<ion-header>
  <draught-header></draught-header>
</ion-header>


<ion-content class="ion-padding">
  <ion-row>
    <ion-col class="ion-text-center"
             size="12" sizeSm="10" offsetSm="1" sizeMd="6" offsetMd="3" sizeLg="6" offsetLg="3" sizeXl="4" offsetXl="4">
      <form [formGroup]="registerForm">
        <h1>Please Register Below</h1>

        <ion-row class="ion-text-center">
          <ion-col class="ion-text-center">
            <ion-item>
              <ion-label color="primary" class="ion-text-left" position="floating">First Name</ion-label>
              <ion-input type="text" formControlName="first_name" tid="firstNameInput"></ion-input>
            </ion-item>
          </ion-col>

          <ion-col class="ion-text-center">
            <ion-item>
              <ion-label color="primary" class="ion-text-left" position="floating">Last Name</ion-label>
              <ion-input type="text" formControlName="last_name" tid="lastNameInput"></ion-input>
            </ion-item>
          </ion-col>
        </ion-row>

        <ion-item>
          <ion-label color="primary" class="ion-text-left" position="floating">Username</ion-label>
          <ion-input type="text" formControlName="username" tid="usernameInput"></ion-input>
        </ion-item>

        <ion-item>
          <ion-label color="primary" class="ion-text-left" position="floating">Email</ion-label>
          <ion-input type="text" formControlName="email" tid="emailInput"></ion-input>
        </ion-item>

        <ion-item>
          <ion-label color="primary" class="ion-text-left" position="floating">Create Password</ion-label>
          <ion-input type="password" formControlName="password" tid="passwordInput"></ion-input>
        </ion-item>

        <ion-item>
          <ion-label color="primary" class="ion-text-left" position="floating">Confirm Password</ion-label>
          <ion-input type="password" formControlName="confirm_password" tid="passwordInput2"></ion-input>
        </ion-item>

        <ion-item>
          <ion-label color="primary" class="ion-text-left" position="floating">Birthday</ion-label>
          <ion-datetime tid="birthDate" (ionChange)="dateSelected()" formControlName="date_of_birth" displayFormat="YYYY/MM/DD" max="{{ maxDate }}"></ion-datetime>
        </ion-item>

        <br>

        <ion-item>
          <ion-label color="primary" class="text-lg-left">Legal Age Confirmation and Disclaimer</ion-label>
        </ion-item>

        <ion-item>
          <ion-label>I am at least 21 years old</ion-label>
          <ion-checkbox formControlName="verify_age" tid="ageVerification" color="primary"></ion-checkbox>
        </ion-item>

        <ion-item>
          <ion-checkbox tid="disclaimerVerification" routerLink="/disclaimer" routerDirection="forward"
                        formControlName="disclaimer_check" color="primary"></ion-checkbox>
          <ion-label>I have read the <a class="underline">disclaimer.</a></ion-label>
        </ion-item>

        <ion-button
                (click)="createUser();"
                tid="registerButton"
                class="ion-margin-top"
                [disabled]="registerForm.invalid">
         Register
        </ion-button>

        <p>Already have an account?</p>
        <a tid="signInLink" routerLink="/sign-in" routerDirection="root">Sign in</a>
      </form>
    </ion-col>
  </ion-row>

</ion-content>
