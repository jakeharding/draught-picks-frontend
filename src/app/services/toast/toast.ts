import { Injectable } from '@angular/core';
import { ToastController } from '@ionic/angular';

/*
  Generated class for the ToastProvider provider.

  See https://angular.io/guide/dependency-injection for more info on providers
  and Angular DI.
*/
@Injectable({
  providedIn: 'root',
})
export class ToastProvider {
  static readonly defaultErrorMsg = 'We are having trouble connecting at the moment. ' +
  'Please check your network connection and try again.';
  private readonly errorToastClass = 'error-toast';
  private readonly successToastClass = 'success-toast';

  constructor(private toastCtrl: ToastController) {}

  errorToast(message: string = ToastProvider.defaultErrorMsg): void {
    this.createToast(message, this.errorToastClass);
  }

  successToast(message: string): void {
    this.createToast(message, this.successToastClass);
  }

  async createToast(message: string,
                    cssClass: string,
                    duration: number = 3000,
                    position: 'top' | 'bottom' | 'middle' = 'top'
              ) {
    const toast = await this.toastCtrl.create({ message, cssClass, duration, position });
    await toast.present();
  }

}
