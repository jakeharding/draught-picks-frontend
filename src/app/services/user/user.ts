import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment as Env } from '../../../environments/environment';
import PageResponse from '../../models/PageResponse';
import User from '../../models/User';
import AuthResponse from '../../models/AuthResponse';
import BeerRating from '../../models/BeerRating';
import Beer from '../../models/Beer';

/**
 * Generated class for the UserProvider provider.
 *
 * See https://angular.io/guide/dependency-injection for more info on providers
 * and Angular DI.
 */
@Injectable({
  providedIn: 'root',
})
export class UserProvider {
  private readonly url: string;
  private readonly resendConfirmEmailUrl: string;
  private readonly confirmEmailUrl: string;
  private readonly passwordResetUrl: string;

  constructor(public http: HttpClient) {
    this.url = `${Env.REST_API_ROOT}/users`;
    this.resendConfirmEmailUrl = `${this.url}/resend-confirm-email`;
    this.confirmEmailUrl = `${this.url}/confirm-email`;
    this.passwordResetUrl = `${this.url}/reset-email`;
  }

  create(userData: AuthResponse | Beer | BeerRating | User) {
    return this.http.post(this.url, userData).toPromise();
  }

  retrieve() {
    return this.http.get(this.url).toPromise().then(({ results }: PageResponse<User>) => {
      return results[0];
    });
  }

  update(user: User): Promise<any> {
    return this.http.put(`${this.url}/${user.uuid}`, user).toPromise();
  }

  resendConfirmEmail({email}): Promise<object> {
    return this.http.post(this.resendConfirmEmailUrl, {email}).toPromise();
  }

  confirmEmail({key}): Promise<any> {
    return this.http.put(this.confirmEmailUrl, {confirm_key: key}).toPromise();
  }

  sendPasswordResetEmail(emailAddress: string): Promise<any> {
    return this.http.post(this.passwordResetUrl, {email: emailAddress}).toPromise();
  }
}
