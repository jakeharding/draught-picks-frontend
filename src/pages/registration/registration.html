<!--
  Generated template for the RegistrationPage page.

  See http://ionicframework.com/docs/components/#navigation for more info on
  Ionic pages and navigation.
-->
<ion-header>
  <ion-navbar color="primary">
    <ion-title text-center>Draught Picks</ion-title>
  </ion-navbar>
</ion-header>


<ion-content padding>
  <ion-row>
    <ion-col text-center col-12 col-sm-10 offset-sm-1 col-md-6 offset-md-3 col-lg-6 offset-lg-3 col-xl-4 offset-xl-4>

      <h1>Please Register Below</h1>
      <ion-label color="primary" text-left stacked>Name</ion-label>

      <ion-row text-center>
        <ion-col text-center>
          <ion-item>
            <ion-input type="text" placeholder="First Name"></ion-input>
          </ion-item>
        </ion-col>

        <ion-col text-center>
          <ion-item>
            <ion-input type="text" placeholder="Last Name"></ion-input>
          </ion-item>
        </ion-col>
      </ion-row>

      <ion-label color="primary" text-left stacked>Email</ion-label>
      <ion-item>
        <ion-input type="text" placeholder="Email"></ion-input>
      </ion-item>

      <ion-label color="primary" text-left stacked>Create Password</ion-label>
      <ion-item>
        <ion-input type="password" placeholder="password"></ion-input>
      </ion-item>

      <ion-label color="primary" text-left stacked>Confirm Password</ion-label>
      <ion-item>
        <ion-input type="password" placeholder="password"></ion-input>
      </ion-item>

      <ion-label color="primary" text-left stacked>Birthday</ion-label>
      <ion-item>
        <ion-label>Your          Birthday</ion-label>
        <ion-datetime displayFormat="MM/DD/YYYY" [(ngModel)]="myDate"></ion-datetime>
      </ion-item>

      <ion-label color="primary" text-left stacked>Legal Age Confirmation and Disclaimer</ion-label>
      <ion-item>
        <ion-label>I am at least 21 years old</ion-label>
        <ion-checkbox color="primary"></ion-checkbox>
      </ion-item>
      <ion-item>
        <ion-label>I have read the disclaimer</ion-label>
        <ion-checkbox color="primary"></ion-checkbox>
      </ion-item>

      <button ion-button (click)="showErrorToastWithButton('top');" margin-top="20px">
        Register
      </button>

      <p>Already have an account?</p>
      <a [navPush]="goToSignInPage">Sign in.</a>

    </ion-col>
  </ion-row>

</ion-content>
