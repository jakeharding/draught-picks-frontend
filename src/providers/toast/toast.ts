import { Injectable } from '@angular/core';
import { ToastController } from 'ionic-angular';

/*
  Generated class for the ToastProvider provider.

  See https://angular.io/guide/dependency-injection for more info on providers
  and Angular DI.
*/
@Injectable()
export class ToastProvider {
  private readonly errorToastClass = 'error-toast';
  private readonly successToastClass = 'success-toast';

  constructor(private toastCtrl: ToastController) {}

  errorToast(message: string): void {
    this.createToast(message, this.errorToastClass);
  }

  successToast(message: string): void {
    this.createToast(message, this.successToastClass);
  }

  createToast(message: string,
              cssClass:string,
              duration: number = 3000,
              position: string = 'top'
              ) {
    this.toastCtrl.create({ message, cssClass, duration, position }).present();
  }

}
