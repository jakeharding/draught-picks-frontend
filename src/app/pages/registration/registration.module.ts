import { NgModule } from '@angular/core';
import { IonicModule } from '@ionic/angular';
import { RegistrationPage } from './registration';
// import { ComponentsModule } from '../../components/components.module';
import { RouterModule } from '@angular/router';

@NgModule({
  declarations: [
    RegistrationPage,
  ],
  imports: [
    IonicModule,
    RouterModule.forChild([{path: 'registration', component: RegistrationPage}]),
    // ComponentsModule,
  ],
})
export class RegistrationPageModule {}
