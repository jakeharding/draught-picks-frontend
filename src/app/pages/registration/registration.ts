import { Component } from '@angular/core';
import { NavController } from '@ionic/angular';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { SignInPage } from '../sign-in/sign-in';
import { DisclaimerPage } from '../disclaimer/disclaimer';
import { UserProvider } from '../../services/user/user';
import CheckboxValidator from '../../validators/CheckboxValidator';
import PasswordValidator from '../../validators/PasswordValidator';
import { BasePage } from '../BasePage';
import { ToastProvider } from '../../services/toast/toast';

/**
 * Generated class for the RegistrationPage page.
 *
 * See https://ionicframework.com/docs/components/#navigation for more info on
 * Ionic pages and navigation.
 */

@Component({
  selector: 'page-registration',
  templateUrl: 'registration.html',
  styleUrls: ['./registration.scss']
})
export class RegistrationPage extends BasePage {
  registerForm: FormGroup;
  goToSignInPage: any;
  goToDisclaimerPage: any;
  maxDate: any;
  verifyAge = false;
  MS_IN_21_YEARS = 662709600000;

  /**
   * RegistrationPage constructor
   * Required constructor for the RegistrationPage class, sets the values maxDate,
   * goToSignInPage, goToDisclaimerPage, and registerForm values accordingly
   */
  constructor(public formBuilder: FormBuilder,
              public navCtrl: NavController,
              public toastProvider: ToastProvider,
              public userProvider: UserProvider ) {
    super('registration');
    this.maxDate = new Date(Date.now() - this.MS_IN_21_YEARS).toISOString();
    this.goToSignInPage = SignInPage;
    this.goToDisclaimerPage = DisclaimerPage;
    this.registerForm = formBuilder.group({
      first_name: ['', [Validators.required]],
      last_name: ['', [Validators.required]],
      date_of_birth: ['', [Validators.required]],
      email: ['', [Validators.required, Validators.email]],
      password: ['', [Validators.required, Validators.minLength(4)]],
      confirm_password: ['', [Validators.required]],
      username: ['', [Validators.required]],
      verify_age: [this.verifyAge, [CheckboxValidator.isChecked, Validators.required]],
      disclaimer_check: [undefined, [CheckboxValidator.isChecked, Validators.required]]
    }, {validator: PasswordValidator.matches});

  }

  /**
   * dateSelected function
   * No Parameters
   * Verifies the age when selecting their birth date
   */
  dateSelected() {
    this.registerForm.controls.verify_age.setValue(true);
  }

  /**
   * createUser function
   * No Parameters
   * Creates a new user after the registration form was submitted and stores
   * all of the registration form values in the database
   */
  public async createUser() {
    console.log(this.registerForm.value.date_of_birth.substring(0, 10));
    this.registerForm.value.date_of_birth = this.registerForm.value.date_of_birth.substring(0, 10);
    await this.userProvider.create(this.registerForm.value).then(
      () => {
        this.navCtrl.navigateRoot('/email-sent');
      }, () => {
        this.toastProvider.errorToast();
      }
    );
  }
}
