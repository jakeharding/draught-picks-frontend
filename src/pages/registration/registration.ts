import {Component} from '@angular/core';
import {IonicPage, NavController, NavParams, ToastController} from 'ionic-angular';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import {SignInPage} from "../sign-in/sign-in";
import {UserProvider} from "../../providers/user/user";

/**
 * Generated class for the RegistrationPage page.
 *
 * See https://ionicframework.com/docs/components/#navigation for more info on
 * Ionic pages and navigation.
 */

@IonicPage()
@Component({
    selector: 'page-registration',
    templateUrl: 'registration.html',
})
export class RegistrationPage {
  registerForm: FormGroup;
  goToSignInPage: any;
  constructor(public formBuilder: FormBuilder,
              public navCtrl: NavController,
              public navParams: NavParams,
              public toastCtrl: ToastController,
              public userProvider: UserProvider ){

    this.goToSignInPage = SignInPage;
    this.registerForm = formBuilder.group({
      first_name: ['', Validators.compose([Validators.maxLength(30)])],
      last_name: ['', Validators.compose([Validators.maxLength(30)])],
      date_of_birth: ['', Validators.compose([Validators.maxLength(30)])],
      email: ['', Validators.compose([Validators.maxLength(30)])],
      password: ['', Validators.compose([Validators.maxLength(30)])],
      username: ['', Validators.compose([Validators.maxLength(30)])]
    })

  }

  public createUser(){
    this.userProvider.createUser(this.registerForm.value).subscribe(
      (response) => {
        location.replace("#/sign-in");
      }
    )
  }
  //
  // ionViewDidLoad() {
  //   console.log('ionViewDidLoad RegistrationPage');
  // }

  public showErrorToastWithButton(userProvider: UserProvider, position: string){
    //this.userProvider.subscribe();
    let toast = this.toastCtrl.create({
      message: 'No network connection, try again later',
      duration: 3000,
      position: "top",
      cssClass: "error-toast"
    });
    toast.present();
  }

}
