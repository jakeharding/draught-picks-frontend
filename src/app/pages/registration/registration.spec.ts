import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { RegistrationPage } from './registration';
import { CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { NavController } from '@ionic/angular';
import { FormBuilder, FormGroupDirective } from '@angular/forms';
import { UserProvider } from '../../services/user/user';
import { ToastProvider } from '../../services/toast/toast';
import { mockNavController, mockToastProvider } from '../../../../setup-jest';

/**
 * registration.spec.ts
 *
 * Created by jake
 * Created on 2018-11-20
 *
 * Test the RegistrationPage.
 */

describe('RegistrationPage', () => {
  let fixture: ComponentFixture<RegistrationPage>;
  let component: RegistrationPage;

  const mockUserProvider = {
    create: jest.fn(() => Promise.resolve())
  };

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      schemas: [CUSTOM_ELEMENTS_SCHEMA],
      declarations: [RegistrationPage, FormGroupDirective],
      providers: [
        { provide: NavController, useValue: mockNavController},
        FormBuilder,
        { provide: ToastProvider, useValue: mockToastProvider},
        { provide: UserProvider, useValue: mockUserProvider},
      ]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(RegistrationPage);
    component = fixture.debugElement.componentInstance;
  });

  it('should initialize the RegistrationPage', () => {
    expect(component).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it('should set registerForm.controls.verify_age.value to true', () => {
    component.dateSelected();
    expect(component.registerForm.controls.verify_age.value).toBe(true);
  });

  test('createUser calls navCtrl.setRoot when userProvider.create succeeds', async () => {
    await component.createUser();
    expect(mockUserProvider.create).toHaveBeenCalledTimes(1);
    expect(mockUserProvider.create).toHaveBeenCalledWith(component.registerForm.value);
    expect(mockNavController.navigateRoot).toHaveBeenCalledTimes(1);
    expect(mockNavController.navigateRoot).toHaveBeenCalledWith('/email-sent');
  });

  test('createUser displays a toast when userProvider.create fails', async () => {
    mockUserProvider.create.mockReturnValue(Promise.reject('error'));
    await component.createUser();
    expect(mockNavController.navigateRoot).not.toHaveBeenCalled();
    expect(mockToastProvider.errorToast).toHaveBeenCalledTimes(1);
  });
});
