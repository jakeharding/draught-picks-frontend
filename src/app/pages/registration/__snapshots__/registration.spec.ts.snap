// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationPage should initialize the RegistrationPage 1`] = `
<page-registration
  MS_IN_21_YEARS={[Function Number]}
  formBuilder={[Function FormBuilder]}
  goToDisclaimerPage={[Function Function]}
  goToSignInPage={[Function Function]}
  maxDate={[Function String]}
  navCtrl={[Function Object]}
  registerForm={[Function FormGroup]}
  toastProvider={[Function Object]}
  userProvider={[Function Object]}
  verifyAge="false"
>
  <draught-header /><ion-content
    padding=""
  >
    <ion-row>
      <ion-col
        class="below-header"
        col-12=""
        col-lg-6=""
        col-md-6=""
        col-sm-10=""
        col-xl-4=""
        offset-lg-3=""
        offset-md-3=""
        offset-sm-1=""
        offset-xl-4=""
        text-center=""
      >
        <form>
          <h1>
            Please Register Below
          </h1>
          <ion-row
            text-center=""
          >
            <ion-col
              text-center=""
            >
              <ion-item>
                <ion-label
                  color="primary"
                  floating=""
                  text-left=""
                >
                  First Name
                </ion-label>
                <ion-input
                  formcontrolname="first_name"
                  tid="firstNameInput"
                  type="text"
                />
              </ion-item>
            </ion-col>
            <ion-col
              text-center=""
            >
              <ion-item>
                <ion-label
                  color="primary"
                  floating=""
                  text-left=""
                >
                  Last Name
                </ion-label>
                <ion-input
                  formcontrolname="last_name"
                  tid="lastNameInput"
                  type="text"
                />
              </ion-item>
            </ion-col>
          </ion-row>
          <ion-item>
            <ion-label
              color="primary"
              floating=""
              text-left=""
            >
              Username
            </ion-label>
            <ion-input
              formcontrolname="username"
              tid="usernameInput"
              type="text"
            />
          </ion-item>
          <ion-item>
            <ion-label
              color="primary"
              floating=""
              text-left=""
            >
              Email
            </ion-label>
            <ion-input
              formcontrolname="email"
              tid="emailInput"
              type="text"
            />
          </ion-item>
          <ion-item>
            <ion-label
              color="primary"
              floating=""
              text-left=""
            >
              Create Password
            </ion-label>
            <ion-input
              formcontrolname="password"
              tid="passwordInput"
              type="password"
            />
          </ion-item>
          <ion-item>
            <ion-label
              color="primary"
              floating=""
              text-left=""
            >
              Confirm Password
            </ion-label>
            <ion-input
              formcontrolname="confirm_password"
              tid="passwordInput2"
              type="password"
            />
          </ion-item>
          <ion-item>
            <ion-label
              color="primary"
              floating=""
              text-left=""
            >
              Birthday
            </ion-label>
            <ion-datetime
              displayformat="YYYY/MM/DD"
              formcontrolname="date_of_birth"
              tid="birthDate"
            />
          </ion-item>
          <br />
          <ion-item>
            <ion-label
              color="primary"
              text-lg-left=""
            >
              Legal Age Confirmation and Disclaimer
            </ion-label>
          </ion-item>
          <ion-item>
            <ion-label>
              I am at least 21 years old
            </ion-label>
            <ion-checkbox
              color="primary"
              formcontrolname="verify_age"
              tid="ageVerification"
            />
          </ion-item>
          <ion-item>
            <ion-checkbox
              color="primary"
              formcontrolname="disclaimer_check"
              tid="disclaimerVerification"
            />
            <ion-label>
              I have read the 
              <a
                class="underline"
              >
                disclaimer.
              </a>
            </ion-label>
          </ion-item>
          <button
            ion-button=""
            margin-top="20px"
            tid="registerButton"
          >
             Register 
          </button>
          <p>
            Already have an account?
          </p>
          <a
            href="#/sign-in"
            tid="signInLink"
          >
            Sign in.
          </a>
        </form>
      </ion-col>
    </ion-row>
  </ion-content>
</page-registration>
`;
